libraries/lib-snapping/SnapUtils.cpp

```diff
@@ -246,142 +250,100 @@ namespace
 }
 
+class ConstantMultiplierSnapItem final : public SnapRegistryItem
+{
+public:
+   ConstantMultiplierSnapItem(
+      const Identifier& internalName, const TranslatableString& label,
+      double multiplier)
+       : SnapRegistryItem { internalName, label }
+       , mMultiplier { multiplier }
+   {
+      assert(mMultiplier > 0.0);
+   }
 
-SnapFunctor SnapToTime(double multiplier) {
-   return [multiplier](SnapConfig, double value, bool nearest)
-   { return SnapWithMultiplier(value, multiplier, nearest); };
-}
+   SnapResult
+   Snap(const AudacityProject&, double time, bool nearest) const override
+   {      
+      return SnapWithMultiplier(time, mMultiplier, nearest);
+   }
 
-SnapResult SnapToSamples (SnapConfig cfg, double value, bool nearest)
-{
-   return SnapWithMultiplier(value, cfg.rate, nearest);
-}
+   SnapResult SingleStep(
+      const AudacityProject& project, double time, bool upwards) const override
+   {
+      const auto step = (upwards ? 1.0 : -1.0) / mMultiplier;
+      const double result = time + step;
 
-SnapFunctor SnapToFrames(double fps)
-{
-   return [fps](SnapConfig, double value, bool nearest)
-   { return SnapWithMultiplier(value, fps, nearest); };
-}
+      if (result < 0.0)
+         return { 0.0, false };
+      
+      return SnapWithMultiplier(result, mMultiplier, true);
+   }
 
-/*
-   bps = tempo / 60
-   1/lower takes 1/bps
-*/
+private:
+   const double mMultiplier;
+};
 
-SnapFunctor SnapToBar ()
+class ProjectDependentMultiplierSnapItem final : public SnapRegistryItem
 {
-   return [](SnapConfig cfg, double value, bool nearest)
+public:
+   ProjectDependentMultiplierSnapItem(
+      const Identifier& internalName, const TranslatableString& label,
+      MultiplierFunctor functor)
+       : SnapRegistryItem { internalName, label }
+       , mMultiplierFunctor { std::move(functor) }
    {
-      // DV: For now, BPM uses quarter notes, i. e. 1/4 = BPM in musical notation
-      const auto quarterDuration = 60.0 / cfg.tempo;
-      const auto beatDuration = quarterDuration * 4.0 / cfg.timeSignature.second;
-      const auto barDuration = beatDuration * cfg.timeSignature.first;
-      const auto multiplier = 1 / barDuration;
-      
-      return SnapWithMultiplier(value, multiplier, nearest);
-   };
-}
+      assert(mMultiplierFunctor);
+   }
 
-SnapFunctor SnapToBeat(int divisor)
-{
-   return [divisor](SnapConfig cfg, double value, bool nearest)
+   SnapResult
+   Snap(const AudacityProject& project, double time, bool nearest) const override
    {
-      const auto quarterDuration = 60.0 / cfg.tempo;
-      // DV: It was decided that for the time being,
-      // BPM sets the duration for quarter notes.
-      // For this reason, `cfg.timeSignature.second` is ignored
-      const auto fracDuration = quarterDuration * 4.0 / divisor;
-      const auto multiplier = 1.0 / fracDuration;
-
-      return SnapWithMultiplier(value, multiplier, nearest);
-   };
-}
+      if (!mMultiplierFunctor)
```

Comment:
Assertion in the constructor, and constancy of the member, should make the null check unnecessary

Reply:
We have very different view on assertions. There are very little cases where I would agree that assertion without a corresponding `if` is not a programming mistake.

Reply:
If that is your opinion, then make the other  constructor consistent.

Reply:
I meant to say, the other unchecked use of mMultiplierFunctor is inconsistent

Reply:
Great catch, sorry for missing it in the first place

Reply:
I don’t like to review a new assertion that has zero comment with it.  The minimum here would be a ‘@pre’ at the declaration of the constructor.